registry/remote/repository.go

```diff
@@ -944,46 +945,99 @@ func (s *manifestStore) FetchReference(ctx context.Context, reference string) (d
-		return ocispec.Descriptor{}, fmt.Errorf("%s %q: invalid response Content-Type: %w", resp.Request.Method, resp.Request.URL, err)
+		return ocispec.Descriptor{}, fmt.Errorf(
+			"%s %q: invalid response `Content-Type` header; %w",
+			resp.Request.Method,
+			resp.Request.URL,
+			err,
+		)
 	}
 
+	// 2. Validate Size
 	size := resp.ContentLength
 	if size == -1 {
-		return ocispec.Descriptor{}, fmt.Errorf("%s %q: unknown response Content-Length", resp.Request.Method, resp.Request.URL)
-	}
-
-	// validate digest if ref is a digest
-	if refDigest, err := ref.Digest(); err == nil {
-		if err = verifyContentDigest(resp, refDigest); err != nil {
-			return ocispec.Descriptor{}, err
+		return ocispec.Descriptor{}, fmt.Errorf(
+			"%s %q: unknown response Content-Length",
+			resp.Request.Method,
+			resp.Request.URL,
+		)
+	}
+
+	// 3. Validate Digest
+	var dgst, respDigest, refDigest, hdrDigest digest.Digest
+	var refDigestErr, hdrDigestErr error
+	if refDigest, err = ref.Digest(); err != nil {
+		refDigestErr = fmt.Errorf(
+			"%s %q: invalid reference digest: `%s`; %w",
+			resp.Request.Method,
+			resp.Request.URL,
+			respDigest,
+			err,
+		)
+	}
+	if hdrDigestStr := resp.Header.Get("Docker-Content-Digest"); hdrDigestStr != "" {
+		if hdrDigest, err = digest.Parse(hdrDigestStr); err != nil {
+			hdrDigestErr = fmt.Errorf(
+				"%s %q: invalid response header value `Docker-Content-Digest`: `%s`; %w",
+				resp.Request.Method,
+				resp.Request.URL,
+				hdrDigestStr,
+				err,
+			)
 		}
-		return ocispec.Descriptor{
-			MediaType: mediaType,
-			Digest:    refDigest,
-			Size:      size,
-		}, nil
-	}
-
-	digestStr := resp.Header.Get("Docker-Content-Digest")
-	if digestStr == "" {
-		return ocispec.Descriptor{}, fmt.Errorf("%s %q: empty response Docker-Content-Digest", resp.Request.Method, resp.Request.URL)
 	}
 
-	contentDigest, err := digest.Parse(digestStr)
-	if err != nil {
-		return ocispec.Descriptor{}, fmt.Errorf("%s %q: invalid response Docker-Content-Digest: %s", resp.Request.Method, resp.Request.URL, digestStr)
+	switch {
+	case calculateDigest:
+		if respDigest, err = calculateDigestFromResponse(resp); err != nil {
+			return ocispec.Descriptor{}, fmt.Errorf("failed to calculate digest on response body; %w", err)
+		}
+		var digests = map[string]digest.Digest{
+			"server-supplied": hdrDigest, // via response header `Docker-Content-Digest`
+			"client-supplied": refDigest, // via request reference string
+		}
+		if err = verifyContentDigests(respDigest, digests); err != nil {
+			return ocispec.Descriptor{}, fmt.Errorf("digest validation failure; %w", err)
+		}
+		dgst = respDigest
+	case refDigestErr == nil:
+		dgst = refDigest
+	case hdrDigestErr == nil:
+		dgst = hdrDigest
 	}
 
 	return ocispec.Descriptor{
 		MediaType: mediaType,
-		Digest:    contentDigest,
+		Digest:    dgst,
 		Size:      size,
 	}, nil
 }
 
+// calculateDigestFromResponse calculates the actual digest of the response body, taking care not to destroy it in the
```

Comment:
nit: Can we limit the length of one line of comment to 80 characters? It's the current convention of the oras-go repo.
We will need to have a contribution guide to call out all the conventions. /cc @shizhMSFT

Reply:
Oh, I've been using line length of 120.  Should I make that change for all comments in the file then?